salt/minion.py

```diff
@@ -587,8 +587,21 @@ def eval_master(self,
                 opts.update(prep_ip_port(opts))
                 opts.update(resolve_dns(opts))
                 try:
-                    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
-                    yield pub_channel.connect()
+                    if self.opts['transport'] == 'detect':
+                        detect_connect = False
+                        self.opts['detect_mode'] = True
+                        for trans in ('tcp', 'zeromq'):
+                            if detect_connect is True:
+                                continue
+                            self.opts['transport'] = trans
+                            pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
+                            yield pub_channel.connect()
+                            if not pub_channel.auth.authenticated:
+                                continue
+                            detect_connect = True
```

Comment:
The `detect_connect` flag looks redundant. The code will be better readable if you'll just break if `authenticated`.